### DISCUSSION TITLE
Understanding `Systems` & the Bevy Runtime

### DISCUSSION THREAD

**Discussion by jdc-pub** [bevyengine/bevy#8956] [Q&A] [answered]
I'm a very early beginner with `bevy`; congratulations on the excellent work here! Reading through examples, it seems like you can define arguments to your `system` functions in any order. I also understand that all types must be unique. For example...

```rust
//
// From the official Bevy examples...
// 

fn greet_people(
    time: Res<Time>, mut timer: ResMut<GreetTimer>, query: Query<&Name, With<Person>>) {
    // update our timer with the time elapsed since the last update
    // if that caused the timer to finish, we say hello to everyone
    if timer.0.tick(time.delta()).just_finished() {
        for name in &query {
            println!("hello {}!", name.0);
        }
    }
}
```

Originally `greet_people` included only the `query` parameter. Ultimately the `time` parameter was added to prevent the function from running too often. Either way, the function was added to the `App` by `.add_system(greet_people)`.

**How does `bevy` know what arguments to pass to each system**?

**Comment by MrGVSV** [comment, marked as answer]
This [post](https://promethia-27.github.io/dependency_injection_like_bevy_from_scratch/introductions.html) by @PROMETHIA-27 might help shed some light. 🙂
